llvm/lib/Transforms/Vectorize/VPlan.h

```diff
@@ -556,6 +556,26 @@ class VPBlockBase {
     return getEnclosingBlockWithPredecessors()->getSinglePredecessor();
   }
 
+  /// This function replaces one predecessor with another, useful when
+  /// trying to replace an old block in the CFG with a new one.
+  void replacePredecessor(VPBlockBase *Old, VPBlockBase *New) {
+    auto I = find(Predecessors, Old);
+    assert(I != Predecessors.end());
+    assert(Old->getParent() == New->getParent() &&
+           "replaced predecessor must have the same parent");
+    *I = New;
```

Comment:
I think this may  leaves `Old` in an invalid state, as we remove it from the list of predecessors here, but don't remove `this` from the `Successors` of `Old`. Can we also remove it from there? Otherwise the callers of the API must either remove the node itself  or completely delete the block

Reply:
Hmm, but if we go down that route then surely we should also care about setting the successor of `New` to `this` as well so that we leave it in a valid state? I guess if this function was private then it wouldn't be part of the interface so it wouldn't matter as much. If we do the extra work here as you suggest, then this patch becomes less about efficiency and more about retaining ordering, which is fine but then I probably ought to change the title!

Reply:
For example, `removeSuccessor` does not leave the `Successor` argument in a valid state either because `this` will still be a predecessor of `Successor`. How about I make the functions private and then we can behave like `removeSuccessor`, unless you see value in these new interfaces being public?

Reply:
Right, let me take a look tomorrow

Reply:
Sure, no problem. I just added a new commit as I realised the the public `reassociateBlocks` interface was leaving `Old` and `New` in an invalid state. Hopefully, this now addresses your concern!

Reply:
Thanks this was what I was about to suggest, as this ensures that `reassociateBlocks` leaves the CFG in a valid state.